crates/pop-cli/src/commands/install/frontend.rs

```diff
@@ -0,0 +1,162 @@
+	let install_cmd = format!(r#"source "{}" && nvm install node"#, nvm_script);
+	cmd("bash", vec!["-c", &install_cmd]).run()?;
+	Ok(())
+}
+
+/// Ensure Bun exists and return the absolute path to the `bun` binary.
+///
+/// # Arguments
+/// * `skip_confirm`: If true, skip user confirmation prompts.
+/// * `cli`: Command line interface.
+pub async fn ensure_bun(skip_confirm: bool, cli: &mut impl cli::traits::Cli) -> Result<PathBuf> {
+	if let Some(path) = which("bun") {
+		return Ok(PathBuf::from(path));
+	}
+	if !skip_confirm &&
+		!cli.confirm(
+			"📦 Do you want to proceed with the installation of the following package: bun ?",
+		)
+		.initial_value(true)
+		.interact()?
+	{
+		return Err(anyhow!("🚫 You have cancelled the installation process."));
+	}
+	// Install Bun using the official installer script
+	run_external_script("https://bun.sh/install", &[]).await?;
+
+	// Try to locate bun in PATH again
+	if let Some(path) = which("bun") {
+		return Ok(PathBuf::from(path));
+	}
+	// Use the default install location from the official installer if not found in PATH
+	let home = std::env::var("HOME").map_err(|_| anyhow!("HOME not set"))?;
+	let bun_abs = PathBuf::from(format!("{home}/.bun/bin/bun"));
+
+	if !bun_abs.exists() {
+		return Err(anyhow!(format!(
+			"Bun installed but not found at {}. Open a new shell or add it to PATH.",
+			bun_abs.display()
+		)));
+	}
+	Ok(bun_abs)
+}
+
+/// Require `npx` to be available.
+pub fn ensure_npx() -> Result<()> {
+	if !has("npx") && !has("npm") {
+		return Err(anyhow!(
+			"`npx` (or npm with npx) not found on PATH. Install NodeJS from https://nodejs.org and re-run."
+		));
+	}
+	Ok(())
+}
+
+fn has(bin: &str) -> bool {
+	cmd("which", vec![bin]).read().is_ok()
+}
+
+fn which(bin: &str) -> Option<String> {
+	cmd("which", vec![bin]).read().ok().map(|s| s.trim().to_string())
+}
+
+/// Install nvm (Node Version Manager) if not already present.
+/// Returns the path to the nvm.sh script.
+async fn install_nvm(cli: &mut impl cli::traits::Cli) -> Result<String> {
+	let home = std::env::var("HOME").map_err(|_| anyhow!("HOME not set"))?;
+	let nvm_script = format!("{home}/.nvm/nvm.sh");
+
+	let nvm_is_installed = std::path::Path::new(&nvm_script).exists();
+	if nvm_is_installed {
+		cli.info("ℹ️ nvm already installed.".to_string())?;
+	} else {
+		run_external_script(NVM_INSTALL_SCRIPT, &[]).await?;
+	}
+	Ok(nvm_script)
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::cli::MockCli;
+
+	#[test]
+	fn has_works() {
+		assert!(has("sh"));
+		assert!(!has("this_binary_should_not_exist_12345"));
+	}
+
+	#[test]
+	fn which_works() {
+		let result = which("sh");
+		assert!(result.is_some() || result.is_none()); // Just verify no panic
+		let result = which("this_binary_should_not_exist_12345");
+		assert!(result.is_none());
+	}
+
+	#[tokio::test]
+	async fn ensure_bun_returns_path_when_bun_exists() {
+		// Only test if bun exists: check we get a path
+		if which("bun").is_some() {
```

Comment:
This is a conditional test. Tests should not rely on the system having a binary or not.